File: generate-business-card.js
import puppeteer from 'puppeteer';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const logoB64 = fs.readFileSync(path.join(__dirname, 'logo.png')).toString('base64');
const fontB64 = fs.readFileSync(path.join(__dirname, 'fonts/InterVariable.woff2')).toString('base64');

// 2" × 3.5" at 96 CSS px/in = 192 × 336px
// deviceScaleFactor 6.25 → 1200 × 2100px output = 600 DPI
const CSS_W = 192;
const CSS_H = 336;
const SCALE = 6.25;

const baseCSS = `
  @font-face {
    font-family: 'Inter';
    src: url('data:font/woff2;base64,${fontB64}') format('woff2-variations');
    font-weight: 100 900;
    font-style: normal;
  }
  *, *::before, *::after { box-sizing: border-box; margin: 0; padding: 0; }
  html, body {
    width: 2in;
    height: 3.5in;
    overflow: hidden;
    background: #000;
    color: #fff;
    font-family: 'Inter', sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }
`;

const frontHTML = `<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8" />
<style>
${baseCSS}

body {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  padding: 0.28in 0.22in 0.22in;
}

.logo {
  width: 0.78in;
  height: 0.78in;
  display: block;
}

.wordmark {
  text-align: center;
  font-size: 0.155in;
  font-weight: 800;
  letter-spacing: 0.18em;
  text-transform: uppercase;
  line-height: 1.25;
  padding-right: 0.155in; /* optical compensation for letter-spacing trailing gap */
}
</style>
</head>
<body>
  <img class="logo" src="data:image/png;base64,${logoB64}" alt="" />
  <div class="wordmark">VERTICAL<br>RECTANGLE</div>
</body>
</html>`;

const backHTML = `<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8" />
<style>
${baseCSS}

body {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 0;
  padding: 0 0.22in;
}

.name {
  font-size: 0.21in;
  font-weight: 800;
  letter-spacing: 0.12em;
  text-transform: uppercase;
  text-align: center;
  padding-right: 0.21in; /* trailing letter-spacing compensation */
}

.artist {
  font-size: 0.13in;
  font-weight: 400;
  letter-spacing: 0.04em;
  text-transform: lowercase;
  text-align: center;
  color: rgba(255,255,255,0.75);
  margin-top: 0.09in;
}

.email {
  font-size: 0.10in;
  font-weight: 400;
  letter-spacing: 0.03em;
  text-align: center;
  color: rgba(255,255,255,0.5);
  margin-top: 0.18in;
}
</style>
</head>
<body>
  <div class="name">Alexis Lucio</div>
  <div class="artist">epsilver</div>
  <div class="email">alexis@verticalrectangle.com</div>
</body>
</html>`;

async function renderCard(html, baseName) {
  const browser = await puppeteer.launch({ args: ['--no-sandbox'] });
  const page = await browser.newPage();

  await page.setViewport({ width: CSS_W, height: CSS_H, deviceScaleFactor: SCALE });
  await page.setContent(html, { waitUntil: 'networkidle0' });

  // PNG — 600 DPI (1200×2100px)
  await page.screenshot({
    path: path.join(__dirname, `${baseName}.png`),
    omitBackground: false,
  });
  console.log(`✓  ${baseName}.png  (1200×2100, 600 DPI)`);

  // PDF — vector text, exact card dimensions
  await page.pdf({
    path: path.join(__dirname, `${baseName}.pdf`),
    width: '2in',
    height: '3.5in',
    printBackground: true,
    margin: { top: 0, right: 0, bottom: 0, left: 0 },
  });
  console.log(`✓  ${baseName}.pdf  (2×3.5 in, print-ready)`);

  await browser.close();
}

(async () => {
  await renderCard(frontHTML, 'card-front');
  await renderCard(backHTML, 'card-back');
  console.log('\nDone. Upload the PDFs to Staples.');
})();

File: generate-business-card.py
import base64
from pathlib import Path
from playwright.sync_api import sync_playwright

ROOT = Path(__file__).parent

logo_b64 = base64.b64encode((ROOT / "logo.png").read_bytes()).decode()
font_var_b64   = base64.b64encode((ROOT / "fonts/InterVariable.woff2").read_bytes()).decode()
font_black_b64 = base64.b64encode((ROOT / "fonts/Inter-Black.otf").read_bytes()).decode()

SCALE = 6.25  # 600 DPI

def base_css(w: str, h: str) -> str:
    return f"""
@font-face {{
  font-family: 'Inter';
  src: url('data:font/woff2;base64,{font_var_b64}') format('woff2-variations');
  font-weight: 100 900;
  font-style: normal;
}}
@font-face {{
  font-family: 'InterBlack';
  src: url('data:font/otf;base64,{font_black_b64}') format('opentype');
  font-weight: 900;
  font-style: normal;
}}
*, *::before, *::after {{ box-sizing: border-box; margin: 0; padding: 0; }}
html, body {{
  width: {w};
  height: {h};
  overflow: hidden;
  background: #000;
  color: #fff;
  font-family: 'Inter', sans-serif;
  -webkit-font-smoothing: antialiased;
}}
"""

# ── FRONT: landscape 3.5" × 2" (content rotated to portrait) ─────────────────
FRONT_HTML = f"""<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8" />
<style>
{base_css("3.5in", "2in")}

body {{
  display: flex;
  align-items: center;
  justify-content: center;
}}

.card-inner {{
  width: 2in;
  height: 3.5in;
  flex-shrink: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  padding: 0.28in 0.22in 0.22in;
  transform: rotate(-90deg);
}}

.logo {{
  width: 0.78in;
  height: 0.78in;
  display: block;
}}

.wordmark {{
  text-align: center;
  font-size: 0.22in;
  font-weight: 900;
  letter-spacing: -0.01em;
  text-transform: uppercase;
  line-height: 1;
}}
</style>
</head>
<body>
  <div class="card-inner">
    <img class="logo" src="data:image/png;base64,{logo_b64}" alt="" />
    <div class="wordmark">VERTICAL<br>RECTANGLE</div>
  </div>
</body>
</html>"""

# ── BACK: landscape 3.5" × 2" ────────────────────────────────────────────────
BACK_HTML = f"""<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8" />
<style>
{base_css("3.5in", "2in")}

body {{
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 0 0.3in;
}}

.name {{
  font-family: 'InterBlack', sans-serif;
  font-size: 0.28in;
  font-weight: 900;
  letter-spacing: -0.01em;
  text-transform: uppercase;
  text-align: center;
  line-height: 1;
}}

.artist {{
  font-family: 'InterBlack', sans-serif;
  font-size: 0.13in;
  font-weight: 900;
  letter-spacing: -0.01em;
  text-align: center;
  color: #fff;
  margin-top: 0.08in;
  line-height: 1;
}}

.name + .name {{
  font-size: 0.196in;
}}

.email {{
  font-family: 'InterBlack', sans-serif;
  font-size: 0.1in;
  font-weight: 900;
  letter-spacing: -0.01em;
  text-align: center;
  color: #fff;
  margin-top: 0.04in;
  line-height: 1;
}}
</style>
</head>
<body>
  <div class="name">Alexis Lucio</div>
  <div class="name">CEO</div>
  <div class="email">alexis@verticalrectangle.com</div>
  <div class="email">verticalrectangle.com</div>
  <div class="email" style="display:inline-flex;align-items:center;gap:0.03in;"><svg viewBox="0 0 24 24" style="width:0.1in;height:0.1in;fill:white;flex-shrink:0;"><path d="M18.244 2.25h3.308l-7.227 8.26 8.502 11.24H16.17l-4.714-6.231-5.401 6.231H2.747l7.73-8.835L1.254 2.25H8.08l4.713 6.231zm-1.161 17.52h1.833L7.084 4.126H5.117z"/></svg> @epsilver_</div>
</body>
</html>"""


def render_card(playwright, html: str, base_name: str, w_in: str, h_in: str):
    # CSS px at 96px/in
    css_w = round(float(w_in.replace("in", "")) * 96)
    css_h = round(float(h_in.replace("in", "")) * 96)
    px_w = round(css_w * SCALE)
    px_h = round(css_h * SCALE)

    browser = playwright.chromium.launch()
    page = browser.new_page(
        viewport={"width": css_w, "height": css_h},
        device_scale_factor=SCALE,
    )
    page.set_content(html, wait_until="networkidle")

    png_path = str(ROOT / f"{base_name}.png")
    page.screenshot(path=png_path, omit_background=False)
    print(f"✓  {base_name}.png  ({px_w}×{px_h}, 600 DPI)")

    pdf_path = str(ROOT / f"{base_name}.pdf")
    page.pdf(
        path=pdf_path,
        width=w_in,
        height=h_in,
        print_background=True,
        margin={"top": "0", "right": "0", "bottom": "0", "left": "0"},
    )
    print(f"✓  {base_name}.pdf  ({w_in}×{h_in}, print-ready)")

    browser.close()


with sync_playwright() as pw:
    render_card(pw, FRONT_HTML, "card-front", "3.5in", "2in")
    render_card(pw, BACK_HTML,  "card-back",  "3.5in", "2in")
    print("\nDone. Upload the PDFs to Staples.")
